src/user_workspaces_server/controllers/resources/abstract_resource.py

```diff
@@ -9,38 +17,48 @@ def __init__(self, config, resource_storage, resource_user_authentication):
         self.passthrough_domain = config.get("passthrough_domain", "")
 
     @abstractmethod
-    def translate_status(self, status):
+    def translate_status(self, status: str) -> Job.Status:
         # Should take a string and translate it into local terminology.
         pass
 
     @abstractmethod
-    def launch_job(self, job, workspace, resource_options):
+    def launch_job(self, job: AbstractJob, workspace: Workspace, resource_options: dict) -> int:
         # Should return resource_job_id
         pass
 
     @abstractmethod
-    def get_resource_job(self, job):
+    def get_resource_job(self, job) -> dict:
         # Should get the resource's job information
         pass
 
     @abstractmethod
-    def get_job_core_hours(self, job):
+    def get_job_core_hours(self, job: Job) -> int:
         # Should return time in hours
         pass
 
     @abstractmethod
-    def stop_job(self, job):
+    def stop_job(self, job: Job) -> bool:
         # Should stop the job on the resource
         pass
 
-    @abstractmethod
-    def validate_options(self, resource_options):
-        # Should determine whether the requested options are valid for a resource
-        # Might be able to implement this at the abstract level once we've defined
-        #   a data model for resource options.
-        pass
+    def validate_options(self, resource_options: dict) -> bool:
+        validator = ParamValidator()
+        validator.validate(resource_options)
+        if validator.errors:
+            logging.error(f"Validation errors: {validator.errors}")
+            raise ValidationException(f"Invalid resource options found: {validator.errors}")
+        return validator.is_valid
 
-    @abstractmethod
-    def translate_options(self, resource_options):
-        # Should translate the options into a format that can be used by the resource
-        pass
+    def translate_option_name(self, option: str) -> str:
+        return self.config.get("parameter_mapping", {}).get(option)
+
+    def translate_options(self, resource_options: dict) -> dict:
```

Comment:
@gesinaphillips - I'm a little confused about the difference between translate_options and translated_options. Seems to me that they could be combined

Reply:
I went back and forth on this abstraction. It might be premature optimization. The logic in `translated_options` represents functionality that I imagined would be repeated and extended by different resources (e.g. how `SlurmAPIResource` has some additional logic about the `gpu_enabled` parameter, whereas `LocalResource` doesn't have any translation requirements at all). I thought that this pattern would let me a) write the reusable translation logic in a single method while also b) providing `translate_options` as the method where any (or in the case of the AbstractResource/LocalResource, no) additional translation is done.

All that said, `translated_options` should be styled with a leading underscore.

Reply:
The other option that I see would be defining `translate_options` as an abstract method in the parent class and defining it explicitly in both child classes--`SlurmAPIResource.translate_options` would be revised to include the logic currently in `translated_options`, and `LocalResource.translated_options` would either just be the logic in `translated_options` or would just return `{}`.

Reply:
Could we move the code in `translated_options` into `translate_options` in the Abstract class? Then, in `SlurmAPIResource` you can use super() to call the Abstract class's `translate_options` and then do the additional translation that is specific to Slurm?

Reply:
Having thought about why I did it that way initially, I'm not going to do that anymore :sweat_smile: Updated.